flink-filesystems/flink-gs-fs-hadoop/src/main/java/org/apache/flink/fs/gs/writer/GSRecoverableWriterCommitter.java

```diff
@@ -0,0 +1,150 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.fs.gs.writer;
+
+import org.apache.flink.core.fs.RecoverableFsDataOutputStream;
+import org.apache.flink.core.fs.RecoverableWriter;
+import org.apache.flink.fs.gs.GSFileSystemOptions;
+import org.apache.flink.fs.gs.storage.BlobStorage;
+import org.apache.flink.fs.gs.utils.BlobUtils;
+import org.apache.flink.util.Preconditions;
+
+import com.google.cloud.storage.BlobId;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Optional;
+
+/** The committer for the GS recoverable writer. */
+class GSRecoverableWriterCommitter implements RecoverableFsDataOutputStream.Committer {
+
+    /** The underlying blob storage. */
+    private final BlobStorage storage;
+
+    /** The GS file system options. */
+    private final GSFileSystemOptions options;
+
+    /** The recoverable writer instance. */
+    private final GSRecoverableWriter writer;
+
+    /** The recoverable writer state for the commit operation. */
+    private final GSRecoverableWriterState state;
+
+    GSRecoverableWriterCommitter(
+            BlobStorage storage,
+            GSFileSystemOptions options,
+            GSRecoverableWriter writer,
+            GSRecoverableWriterState state) {
+        this.storage = Preconditions.checkNotNull(storage);
+        this.options = Preconditions.checkNotNull(options);
+        this.writer = Preconditions.checkNotNull(writer);
+        this.state = Preconditions.checkNotNull(state);
+    }
+
+    @Override
+    public void commit() throws IOException {
+
+        // compose all the component blob ids into the final blob id. if the component blob ids are
+        // in the same bucket as the final blob id, this can be done directly. otherwise, we must
+        // compose to a new temporary blob id in the same bucket as the component blob ids and
+        // then copy that blob to the final blob location
+        if (state.finalBlobId.getBucket().equals(state.getTemporaryBucketName(options))) {
+
+            // compose directly to final blob
+            composeBlobs(
+                    state.getComponentBlobIds(options),
+                    state.finalBlobId,
+                    options.writerContentType);
+
+        } else {
+
+            // compose to a temporary blob id, then copy to final blob id
+            BlobId intermediateBlobId = state.createTemporaryBlobId(options);
+            composeBlobs(
+                    state.getComponentBlobIds(options),
+                    intermediateBlobId,
+                    options.writerContentType);
+            storage.copy(intermediateBlobId, state.finalBlobId);
+        }
+
+        // clean up after commit
+        writer.cleanupRecoverableState(state);
```

Comment:
I'm not entirely sure about cleaning the temp blobs here.

Ideally, for each `Recoverable` that has been created, there should be a call to `cleanupRecoverableState` that cleans up the corresponding state.

I think it serves as a good safe net that we do a cleanup on committing, if the cleaning doesn't cause any problem. However, as you have mentioned, it is possible that the job restarts from an early checkpoint. I think we should not overwrite an existing committed file from the storage. But what if the file is manually cleaned up and the job is intentionally restarted from an early checkpoint?

Maybe `GSRecoverableWriterCommitter` should not try to be over smart. We might tolerant some un-cleaned temp files, and it should be `StreamingFileSink`'s responsibility to make sure `cleanupRecoverableState` is always called.

Reply:
I wasn't sure exactly when ```cleanupRecoverableState``` was called in the normal flow of things, so I ran a test where I set a breakpoint on that method and wrote data through a StreamingFileSink. What I observed was that ```cleanupRecoverableState``` did not get called as blobs were successfully committed, so no cleanup was occurring. 

From the description of the [method](https://ci.apache.org/projects/flink/flink-docs-master/api/java/org/apache/flink/core/fs/RecoverableWriter.html#cleanupRecoverableState-org.apache.flink.core.fs.RecoverableWriter.ResumeRecoverable-), it seemed plausible to me that this was only called on failure:
> Frees up any resources that were previously occupied in order to be able to recover from a (potential) failure

... but I wasn't sure.

But yes, I agree, it would make more sense to me if ```cleanupRecoverableState``` were called on both successes and failures, and in that case the call in ```commit``` should be removed.

If it's certainly possible that I made a mistake in my testing. Should ```cleanupRecoverabeState``` be getting called after successful commits?

Reply:
I think `cleanupRecoverableState` is not called immediately on blobs committed, but should rather be called on the next success checkpoint.
I've talked to Guowei offline. He is also a Flink committer and one of the main contributor of streaming file sink. According to him, there are some known issues that `cleanupRecoverableState` may not always get called, e.g., the next success checkpoint never happen. Efforts for fixing those issues are still in progress.

Reply:
I see. I don't actually recall seeing ```cleanupRecoverableState``` being called at any point on successful writes, even after the next checkpoint completed, but it's possible I'm wrong about that. I'll run another quick check to make sure I do see ```cleanupRecoverableState``` being called when expected, and if so, I'll remove that call during commit. I'll report back what I find.

Reply:
On second thought, if we cannot clean anything in `cleanupRecoverableState`, it might be our only chance to clean the temporary blobs on committing. This might have a higher priority than supporting manually recover from an early checkpoint.

I think the following issues are closely related and really need to be clarified consistently.
- What are the relationships between snapshots (resumables) and temporary blobs
- Actions for `cleanupRecoverableState`
- Actions for `commit`

Reply:
Yes, agreed. I've set up a few test cases locally and watched how the various methods (```cleanupRecoverableState```, ```commit```, etc.) get called using both a [DefaultRollingPolicy](https://ci.apache.org/projects/flink/flink-docs-stable/api/java/org/apache/flink/streaming/api/functions/sink/filesystem/rollingpolicies/DefaultRollingPolicy.html) and a [CheckpointRollingPolicy](https://ci.apache.org/projects/flink/flink-docs-stable/api/java/org/apache/flink/streaming/api/functions/sink/filesystem/rollingpolicies/CheckpointRollingPolicy.html) with a [StreamingFileSink](https://ci.apache.org/projects/flink/flink-docs-stable/dev/connectors/streamfile_sink.html). I think I see what's going on here -- I'll share my findings with you and please let me know what you think.

First, a class of interest is the [Bucket](https://github.com/apache/flink/blob/master/flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/filesystem/Bucket.java) class, which is written to by a streaming file sink and is responsible for interacting with the underlying storage, via a [BucketWriter](https://ci.apache.org/projects/flink/flink-docs-master/api/java/org/apache/flink/streaming/api/functions/sink/filesystem/BucketWriter.html), applying the appropriate [RollingFilePolicy](https://ci.apache.org/projects/flink/flink-docs-release-1.9/api/java/org/apache/flink/streaming/api/functions/sink/filesystem/RollingPolicy.html). 

Essentially, a bucket accepts elements via ```write``` and writes them to the currently active (or "in progress") writer. The bucket also decides when to "roll" the file, i.e. when to finish writing to the currently in-progress file and start writing to a new one. To keep track of things, internally the bucket maintains two maps:

* [inProgressFileRecoverablesPerCheckpoint](https://github.com/apache/flink/blob/11f5abc41057cf119548cb03dfcc97d755e0029a/flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/filesystem/Bucket.java#L63), a map from checkpoint id to an [InProgressFileRecoverable](https://github.com/apache/flink/blob/11f5abc41057cf119548cb03dfcc97d755e0029a/flink-connectors/flink-file-sink-common/src/main/java/org/apache/flink/streaming/api/functions/sink/filesystem/InProgressFileWriter.java#L57). Conceptually, this holds the recoverable for the in-progress writer -- if there is one -- associated with a given checkpoint, and it's populated here, in in [onReceptionOfCheckpoint](https://github.com/apache/flink/blob/11f5abc41057cf119548cb03dfcc97d755e0029a/flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/filesystem/Bucket.java#L275):

```
        if (inProgressPart != null) {
            inProgressFileRecoverable = inProgressPart.persist();
            inProgressFileCreationTime = inProgressPart.getCreationTime();
            this.inProgressFileRecoverablesPerCheckpoint.put(
                    checkpointId, inProgressFileRecoverable);
        }
    
```

So, recoverables in ```inProgressFileRecoverablesPerCheckpoint``` will result from calls to ```persist```.

* [pendingFileRecoverablesPerCheckpoint](https://github.com/apache/flink/blob/11f5abc41057cf119548cb03dfcc97d755e0029a/flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/filesystem/Bucket.java#L67), a map from checkpoint id to a list of [PendingFileRecoverable](https://github.com/apache/flink/blob/11f5abc41057cf119548cb03dfcc97d755e0029a/flink-connectors/flink-file-sink-common/src/main/java/org/apache/flink/streaming/api/functions/sink/filesystem/InProgressFileWriter.java#L60). Conceptually, this holds all recoverables for "pending files", which means files that are no longer being written to, having been rolled. This happens in [closePartFile](https://github.com/apache/flink/blob/11f5abc41057cf119548cb03dfcc97d755e0029a/flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/filesystem/Bucket.java#L259):

```
        if (inProgressPart != null) {
            pendingFileRecoverable = inProgressPart.closeForCommit();
            pendingFileRecoverablesForCurrentCheckpoint.add(pendingFileRecoverable);
            inProgressPart = null;
        }
```

So, recoverables in ```pendingFileRecoverablesPerCheckpoint``` will result from calls to ```closeForCommit```.

Last, consider what happens in [onSuccessfulCompletionOfCheckpoint](https://github.com/apache/flink/blob/11f5abc41057cf119548cb03dfcc97d755e0029a/flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/filesystem/Bucket.java#L31):

```
    void onSuccessfulCompletionOfCheckpoint(long checkpointId) throws IOException {
        checkNotNull(bucketWriter);

        Iterator<Map.Entry<Long, List<InProgressFileWriter.PendingFileRecoverable>>> it =
                pendingFileRecoverablesPerCheckpoint
                        .headMap(checkpointId, true)
                        .entrySet()
                        .iterator();

        while (it.hasNext()) {
            Map.Entry<Long, List<InProgressFileWriter.PendingFileRecoverable>> entry = it.next();

            for (InProgressFileWriter.PendingFileRecoverable pendingFileRecoverable :
                    entry.getValue()) {
                bucketWriter.recoverPendingFile(pendingFileRecoverable).commit();
            }
            it.remove();
        }

        cleanupInProgressFileRecoverables(checkpointId);
    }
```

... and [cleanupInProgressFileRecoverables](https://github.com/apache/flink/blob/11f5abc41057cf119548cb03dfcc97d755e0029a/flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/filesystem/Bucket.java#L336):

```
    private void cleanupInProgressFileRecoverables(long checkpointId) throws IOException {
        Iterator<Map.Entry<Long, InProgressFileWriter.InProgressFileRecoverable>> it =
                inProgressFileRecoverablesPerCheckpoint
                        .headMap(checkpointId, false)
                        .entrySet()
                        .iterator();

        while (it.hasNext()) {
            final InProgressFileWriter.InProgressFileRecoverable inProgressFileRecoverable =
                    it.next().getValue();

            // this check is redundant, as we only put entries in the
            // inProgressFileRecoverablesPerCheckpoint map
            // list when the requiresCleanupOfInProgressFileRecoverableState() returns true, but
            // having it makes
            // the code more readable.

            final boolean successfullyDeleted =
                    bucketWriter.cleanupInProgressFileRecoverable(inProgressFileRecoverable);
            if (LOG.isDebugEnabled() && successfullyDeleted) {
                LOG.debug(
                        "Subtask {} successfully deleted incomplete part for bucket id={}.",
                        subtaskIndex,
                        bucketId);
            }
            it.remove();
        }
    }```
```

First, any pending recoverables (meaning ones resulting from calls to ```closeForCommit```, for rolled files) are committed. This is done for the current checkpoint id and any prior checkpoint ids. Then, any in-progress recoverables (meaning ones resulting from calls to ```persist```, for the then in-progress writer) are cleaned up. This is done for any checkpoints prior to the current one. Importantly, recoverables that result from calls to ```closeForCommit``` do not seem to be cleaned up; only the in-progress ones do.

I had been using a ```CheckpointRollingPolicy``` in my tests, and I never saw ```cleanupRecoverableState``` being called on my recoverable writer, even after letting many checkpoints complete. This makes sense to me now -- since the policy I was using closed any in-progress writers on checkpoints, there was never an in-progress writer when the post-checkpoint handling ran, and so that cleanup didn't occur.

I tried using a default rolling policy, set to roll faster than the checkpoint interval, and in that case, I did see ```cleanupRecoverableState``` get called, but only for the ones that were in progress at the time of the checkpoint (i.e. not the ones that had rolled).

So, this suggests to me that it may be expected that ```commit``` perform its own cleanup, even though I don't see that explicitly stated anywhere. Unless we think this is a bug and that ```cleanupRecoverableState``` should be being called after successful commits ...

In any case, you're absolutely right that ```cleanupRecoverableState``` can get called along the way for an active recoverable write, and so we can't delete anything there that we might need later.  So my current implementation is wrong. How to fix depends on how we resolve the commit/cleanup issue:
* If we think that ```commit``` should perform its own cleanup, then I could remove the cleanup code from ```cleanupRecoverableState``` and put it somewhere where ```commit``` can call it.
* If we think that ```cleanupRecoverableState``` should be being called for the commit recoverables, and we feel we can change the existing behavior, then we could leave the code where it is but only perform the cleanup if the supplied recoverable is, in fact, a *commit* recoverable (not a *resume* recoverable)

Either way, I'd think we could keep the existing cleanup algorithm, which deletes for non-orphaned and orphaned temporary blobs, based on the name prefix, i.e. if it's safe to delete the non-orphaned temporary blobs, it should be safe to delete orphaned ones, too.

Last, you said:

> This might have a higher priority than supporting manually recover from an early checkpoint.

The conflict you're referring to -- not being able to manually recover from an early checkpoint after a commit if the commit deletes the temporary blobs -- seems like it would be a problem for *any* cleanup scheme where the commit operation, to succeed, strictly requires the temporary files to be present. 

It was actually this sort of scenario that made me consider an alternate way of handling commit, before, which was to compose the final blob from the temporary blobs *only if it didn't already exist*. My thought was that, if we assume idempotency, it should be ok to skip recreating the final blob if it already exists. And this would allow restoring from earlier checkpoints without failure even after the temporary blobs were deleted on commit. 

I just thought I'd mention that again in case it could help with this issue ...

Reply:
One more thought on recovering from earlier checkpoints.

Another way to make this possible would be not to do any cleanup in ```commit``` or ```cleanupRecoverableState``` and leave the responsibility for cleanup of temporary blobs to the caller. The most obvious way for the caller to do this would be to specify a separate bucket for temporary blobs and then to apply a TTL.

This would make things nice and simple -- temporary files would stay around for the TTL period, and it would be the responsibility of the caller to choose an appropriate TTL to balance recovery and storage needs. The ability to recover wouldn't depend on whether ```commit``` or ```cleanupRecoverableState``` or any other method had been called at some point, it would just depend on the age of the temporary blobs. This actually strikes me as a nice configuration, one that I might like to use.

The downside, of course, would be that if this were the default configuration, then callers who write to GS files without supplying any additional configuration (i.e. no options) would leak temporary blobs. If this were not the default configuration, then we'd be looking at some sort of option to enable this "no cleanup" mode.

Reply:
> How to fix depends on how we resolve the commit/cleanup issue:
> - If we think that commit should perform its own cleanup, then I could remove the cleanup code from cleanupRecoverableState and put it somewhere where commit can call it.
> - If we think that cleanupRecoverableState should be being called for the commit recoverables, and we feel we can change the existing behavior, then we could leave the code where it is but only perform the cleanup if the supplied recoverable is, in fact, a commit recoverable (not a resume recoverable)

I totally agree with your analysis regarding `commit` and `cleanupRecoverableState`. Double-checked with @gaoyunhaii, one of the original contributors of streaming file sink. According to him, `commit` indeed should cleanup the temporary file if any. So my previous comments were incorrect, sorry for the misleading.

> It was actually this sort of scenario that made me consider an alternate way of handling commit, before, which was to compose the final blob from the temporary blobs only if the final blob didn't already exist. My thought was that, if we assume idempotency, it should be ok to skip recreating the final blob if it already exists. And this would allow restoring from earlier checkpoints without failure even after the temporary blobs were deleted on commit.

I'm not entirely sure about the idempotency. Ideally, a file should not be committed twice. IIUC, the only chance a file get committed twice is that, the job fails after committing the file and before the next success checkpoint. In that case, the recovered job does not know whether the file has been successfully committed, and would have to retry committing the file. In such cases, the content of file is guaranteed identical, and `commitAfterRecovery` should be called. Apart from the above case, I think we should fail when a job tries to commit a file twice, which likely indicates something being wrong and there's no guarantee on the idempotency.

To be specific, I think we should take no action for `commitAfterRecovery` and fail for `commit` if the final blob already exist (and/or the temporary blobs do not exist).

> Another way to make this possible would be not to do any cleanup in commit or cleanupRecoverableState and leave the responsibility for cleanup of temporary blobs to the caller. The most obvious way for the caller to do this would be to specify a separate bucket for temporary blobs and then to apply a TTL.

I'd suggest not to rely on the TTL for cleaning up temporary blobs. In addition to what you've mentioned, that this requires additional user configurations, I've some other concerns.
- If the user does not specify a dedicated bucket for the temporary blobs, the temporary blobs will stay as long as the final blobs, since there's no way to specify different TTLs for temporary and final blobs.
- The temporary bucket may take much larger space. A user may specify a relative larger TTL that a small fraction of suspended jobs can be resumed within that time. If the unused temporary files are not cleaned, we would need as large space as to keep all temporary blobs generated during that time, which is undesired.

Reply:
Great -- that all sounds good. Thanks for walking through the failure scenario, that was the scenario I was worried about, i.e. would there be some way to get into a nonrecoverable state. Agreed that, in those scenarios, ```commitAfterRecovery``` should be called and it should allow recovery.

So, to confirm, we'll go with:
* ```commit```
  * If the final blob already exists, fail
  * If the final blob doesn't exist, compose it from the temp blobs (failing if the temp blobs don't exist)
  * Last, delete all temp blobs associated with the final blob, if no failure in prior steps
* ```commitAfterRecovery```
  * If the final blob already exists, do nothing
  * If the final blob doesn't exist, compose it from the temp blobs (failing if the temp blobs don't exist)
  * Last, delete all temp blobs associated with the final blob, if no failure in prior steps
* ```cleanupRecoverableState```
  * Do nothing

If that's right, then I think we have a plan here.

Reply:
Thanks for the summary. Yes, I think we are on the same page.

Reply:
Done in [34ad465](https://github.com/apache/flink/pull/15599/commits/34ad465ca419b0afdb399e53deca8b27027f7191).